df_to_vec.py

```diff
@@ -0,0 +1,242 @@
+            pass
+
+    return vector
+
+
+class Datapoint(ABC):
+    """
+    Abstract class to represent a datapoint
+    """
+
+    @property
+    @abstractmethod
+    def feature_lengths(self) -> Dict[str, int]:
+        """
+        The lengths (number of vectors) of the features
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def feature_types(self) -> Dict[str, str]:
+        """
+        The types (datapoint_type, code or language) of the features
+        """
+        pass
+
+    def vector_length(self) -> int:
+        """
+        The length of the whole vector for this datapoint
+        """
+        return sum(self.feature_lengths.values()) + len(self.feature_lengths.values()) - 1
+
+    def to_vec(self) -> np.ndarray:
+        """
+        The vector for this datapoint
+
+        The vector contains all the features specified in the subclass. Natural language features are converted to
+        vectors using the word2vec models.
+        """
+        datapoint = np.zeros((self.vector_length(), VEC_LENGTH))
+
+        separator = np.ones(VEC_LENGTH)
+
+        position = 0
+        for feature, feature_length in self.feature_lengths.items():
+
+            if self.feature_types[feature] == 'datapoint_type':
+                datapoint[position] = self.datapoint_type_vector()
+                position += 1
+
+            if self.feature_types[feature] == 'code' or self.feature_types[feature] == 'language':
+                vectorized_feature = vectorize_string(
+                    self.__dict__[feature] if isinstance(self.__dict__[feature], str) else '',
+                    feature_length,
+                    w2v_models[self.feature_types[feature]]
+                )
+
+                for word in vectorized_feature:
+                    datapoint[position] = word
+                    position += 1
+
+            if self.feature_types[feature] == 'padding':
+                for i in range(0, feature_length):
+                    datapoint[position] = np.zeros(VEC_LENGTH)
+                    position += 1
+
+            # Add separator after each feature
+            if position < len(datapoint):
+                datapoint[position] = separator
+                position += 1
+
+        return datapoint
+
+    def to_be_predicted_to_vec(self) -> np.ndarray:
+        """
+        A vector representation of what needs to be predicted, in this case the type
+        """
+        vector = np.zeros(NUMBER_OF_TYPES)
+        vector[self.type] = 1
+        return vector
+
+    @abstractmethod
+    def datapoint_type_vector(self) -> np.ndarray:
+        """
+        The vector corresponding to the type
+        """
+        pass
+
+    def __repr__(self) -> str:
+        values = list(map(lambda kv: kv[0] + ': ' + repr(kv[1]), self.__dict__.items()))
+        values = "\n\t" + ",\n\t".join(values) + "\n"
+        return type(self).__name__ + "(%s)" % values
+
+
+class ParameterDatapoint(Datapoint):
```

Comment:
Perhaps adding a docstring here to describe what a Parameter DataPoint consists out of? (Same goes for Return DataPoint)

Reply:
Done!